File: turtlemaster.py
import sys

dr = [-1, 0, 1, 0]
dc = [0, 1, 0, -1]
r, c = 0, 0

grid = []
for x in range(8):
    line = input()
    grid.append(list(line))
    for y in range(8):
        if line[y] == "T":
            r = x
            c = y

dir = 1
line = input()
bug = False;

for move in line:
    ##print(r, " ", c, " ", dir)
    if bug:
        break

    match move:
        case "F": 
            r += dr[dir]
            c += dc[dir]

            if r < 0 or r >= 8 or c < 0 or c >= 8:
                bug = True;
            elif grid[r][c] == "C" or grid[r][c] == "I":
                bug = True;
        case "R":
            dir = (dir + 1) % 4
        case "L":
            dir = (dir + 3) % 4
        case "X":
            tr = r + dr[dir]
            tc = c + dc[dir]
            if tr < 0 or tr >= 8 or tc < 0 or tc >= 8:
                bug = True
            elif grid[tr][tc] == "I":
                grid[tr][tc] = "."
                ##print(grid[tr][tc])
            elif grid[tr][tc] != "D":
                bug = True
if bug:
    print("Bug!")
elif grid[r][c] == "D":
    print("Diamond!")
else:
    print("Bug!")

File: turtlemaster.cpp
#include <iostream>
#include <iomanip>
#include <vector>
#include <unordered_map>

using namespace std;

int main()
{
    vector<int> dr = {-1, 0, 1, 0};
    vector<int> dc = {0, 1, 0, -1};
    int r = 0, c = 0;

    vector<vector<char>> grid(8, vector<char>(8));
    for (int i = 0 ; i < 8 ; ++i)
    {
        for(int j = 0 ; j < 8 ; ++j)
        {
            cin >> grid[i][j];
            if (grid[i][j] == 'T' ) { r = i; c = j; }
        }
    }

    int dir = 1;
    string line, junk;
    getline(cin, junk);
    getline(cin, line);
    for(auto move: line)
    {
        //cout << r << " " << c << " " << dir << '\n';
        switch(move)
        {
            case 'F': 
            {
                r += dr[dir]; 
                c += dc[dir];

                if (r < 0 || r >= 8 || c < 0 || c >= 8) {
                    cout << "Bug!\n"; return 0;
                }

                if (grid[r][c] == 'C' || grid[r][c] == 'I') { 
                    cout << "Bug!\n"; return 0;
                }
                break;
            }
            case 'R': 
            {
                dir = (dir + 1) % 4;
                break;
            }
            case 'L': 
            {
                dir = (dir + 3) % 4;
                break;
            }
            case 'X': 
            {
                int tr = r + dr[dir]; 
                int tc = c + dc[dir]; 
                if(tr < 0 || tc < 0 || tr >= 8 || tc >= 8) {
                    cout << "Bug!\n"; return 0;
                }

                if (grid[tr][tc] == 'I') {
                    grid[tr][tc] = '.';
                } else if (grid[tr][tc] != 'D') {
                    cout << "Bug!\n"; return 0;
                }
                break;
            }
            default: break;
        }
    }

    if (grid[r][c] == 'D' ) { cout << "Diamond!\n"; }
    else { cout << "Bug!\n"; }
}
